### PR TITLE
feat(tainttoleration): return Skip status in PreScore

### LINKED ISSUE #115745: [Umbrella] Skip Score plugins when coupled PreScore plugin returns Skip status
### What would you like to be added?

It's much similar to https://github.com/kubernetes/kubernetes/issues/114399.
In https://github.com/kubernetes/kubernetes/issues/114827, we support Skip in PreScore/Score as well. (Thanks @kidddddddddddddddddddddd)
So, the plugin can return Skip in PreScore when it doesn't do anything in Score in that scheduling cycle.

For example, if the pod doesn't have any preferred inter-pod affinity, the inter-pod affinity plugin PreScore can return Skip, and the scheduler won't run the inter-pod affinity plugin Score.

Let's change our in-tree Score plugins to return Skip.

Here, I just listed all score plugins, **not all of them may not need to return Skip in PreScore though**. Assignee, please make sure your PreScore/Score plugin really needs to return Skip or not.

Please feel free to assign what you'd like to work on. And **please comment in this issue before starting to work on so that we can avoid many people from working on the same thing!**

- [x] ImageLocality @tangwz  no need.
- [x] InterPodAffinity @utam0k https://github.com/kubernetes/kubernetes/pull/117794
- [x] NodeAffinity @sanposhiho https://github.com/kubernetes/kubernetes/pull/117024
- [x] Fit @tangwz no need.
- [x] BalancedAllocation @tangwz no need.
- [x] PodTopologySpread @utam0k https://github.com/kubernetes/kubernetes/pull/118608
- [x] SelectorSpread; no need. It'll soon be removed. https://github.com/kubernetes/kubernetes/issues/108444
- [x] TaintToleration @tangwz no need. https://github.com/kubernetes/kubernetes/pull/118324
- [x] VolumeBinding @kidddddddddddddddddddddd https://github.com/kubernetes/kubernetes/pull/115768

### Why is this needed?

The scheduler can know which plugins it really needs/doesn't need to run to get the scheduling result.
It'd contribute to the scheduling performance.

/sig scheduling
/assign

### PR DESCRIPTION
**Pull request by tangwz**
#### What type of PR is this?

/kind feature

#### What this PR does / why we need it:

This PR make TaintToleration PreScore returning Skip when nothing to do in TaintToleration Score.

#### Which issue(s) this PR fixes:

Fixes #115745

#### Does this PR introduce a user-facing change?

```release-note
None
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

File: pkg/scheduler/framework/plugins/tainttoleration/taint_toleration_test.go [hunk 1]
```diff
@@ -239,6 +255,12 @@ func TestTaintTolerationScore(t *testing.T) {
 
 			p, _ := New(nil, fh)
 			status := p.(framework.PreScorePlugin).PreScore(ctx, state, test.pod, test.nodes)
+			if diff := cmp.Diff(test.wantPreScoreStatus, status); diff != "" {
```
Comment (AxeZhan) [inline pkg/scheduler/framework/plugins/tainttoleration/taint_toleration_test.go:258]: can you change the `reflect.DeepEqual` in line 282 to `cmp.Diff` as well?

**Comment by sanposhiho** [comment]
I think `len(nodes) == 0` will never happen. Could you double-check?
Then, instead of returning Skip, can we just eliminate ``len(nodes) == 0` condition check?

**Comment by tangwz** [comment]
> I think `len(nodes) == 0` will never happen. Could you double-check? Then, instead of returning Skip, can we just eliminate ``len(nodes) == 0` condition check?

Yes, I think so.

So, we don't return Skip here, and I think `ImageLocality` no need too?

**Comment by sanposhiho** [comment]
Yep, agree with that. Let me mark taint toleration and image locality done.

File: pkg/scheduler/framework/plugins/tainttoleration/taint_toleration.go [hunk 2]
```diff
@@ -98,7 +98,7 @@ func getAllTolerationPreferNoSchedule(tolerations []v1.Toleration) (tolerationLi
 // PreScore builds and writes cycle state used by Score and NormalizeScore.
 func (pl *TaintToleration) PreScore(ctx context.Context, cycleState *framework.CycleState, pod *v1.Pod, nodes []*v1.Node) *framework.Status {
 	if len(nodes) == 0 {
-		return nil
+		return framework.NewStatus(framework.Skip)
```
Comment (ahg-g) [inline pkg/scheduler/framework/plugins/tainttoleration/taint_toleration.go:101]: score will not run anyways if there are no nodes to score, so this PR is not necessary.

**Pull request closed without merging**